pkg/k8sops/types.go

```diff
@@ -0,0 +1,128 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package k8sops
+
+import (
+	myspec "github.com/m3db/m3db-operator/pkg/apis/m3dboperator/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type K8sops interface {
+	// StatefulSetName provides a formatted string to use for naming StatefulSets
+	StatefulSetName(clusterName, isolationGroup string) string
+
+	// ListM3DBCluster will list all the CRDS for M3DBClusters in all namespaces
+	ListM3DBCluster() (*myspec.M3DBClusterList, error)
+
+	// GetM3DBCluster will get the M3DBCluster CRD
+	GetM3DBCluster(namespace, name string) (*myspec.M3DBCluster, error)
+
+	// GetCRD will get a CRD
+	GetCRD(name string) (*apiextensionsv1beta1.CustomResourceDefinition, error)
+
+	// CreateCRD checks if M3DB CRD exists. If not, create
+	CreateCRD(name string) error
+
+	// GenerateCRD generates the crd object needed for the M3DBCluster
+	GenerateCRD() *apiextensionsv1beta1.CustomResourceDefinition
+
+	// UpdateCRD will update a CRD
+	UpdateCRD(cluster *myspec.M3DBCluster) (*myspec.M3DBCluster, error)
+
+	// GenerateStatefulSet provides a statefulset object for a m3db cluster
+	GenerateStatefulSet(
+		cluster *myspec.M3DBCluster,
+		clusterSpec myspec.ClusterSpec,
+		svcCfg myspec.ServiceConfiguration,
+		isolationGroup string,
+		instanceAmount *int32,
+	) (*appsv1.StatefulSet, error)
+
+	// GenerateMaps will produce the proper datastructure for v1.Labels
+	GenerateMaps(kind string, svcCfg myspec.ServiceConfiguration) map[string]string
+
+	// GenerateService will produce resource configured according to the spec's
+	// serviceConfiguration fields
+	GenerateService(svcCfg myspec.ServiceConfiguration) *v1.Service
+
+	// GenerateServicePorts will produce the correct ServicePort or ContainerPort
+	// resources
+	GenerateServicePorts(ports []myspec.Port) []v1.ServicePort
+
+	// GenerateContainerPorts will produce ServicePorts given a
+	// ServiceConfiguration
+	GenerateContainerPorts(ports []myspec.Port) []v1.ContainerPort
+
+	// MonitorM3DBEvents watches for new or removed clusters
+	MonitorM3DBEvents(stopchan chan struct{}) (<-chan *myspec.M3DBCluster, <-chan error)
```

Comment:
Go channels work in weird ways (e.g. sending on a blocked channel can cause a panic), it's frowned upon to use channels across interface boundaries. Could you use something like https://github.com/m3db/m3x/blob/master/watch/watch.go to wrap the types being passed around.

It'll give you much clearer expectations about ownership - who has to close which object and so on.

Reply:
Yea ended up moving this within the Controller package and handling the channels within the package itself